### ISSUE TITLE
The index_parallel task possibly fails for taskId conflict

### ISSUE THREAD

**Issue by yuanlihan** [apache/druid#9958]
### Affected Version

0.16.x and later.

### Description
The `index_parallel` task possibly fails for taskId conflict of `single_phase_sub` tasks. Recently I got failures of `index_parallel` tasks caused by taskId conflict with the following error:
```
2020-05-29T15:53:24,994 ERROR [task-runner-0-priority-0] org.apache.druid.indexing.overlord.SingleTaskBackgroundRunner - Exception while running task[AbstractTask{id='index_parallel_ds_name_rollup_for_pv_2020-05-29T07:52:45.317Z', groupId='index_parallel_ds_name_rollup_for_pv_2020-05-29T07:52:45.317Z', taskResource=TaskResource{availabilityGroup='index_parallel_ds_name_rollup_for_pv_2020-05-29T07:52:45.317Z', requiredCapacity=1}, dataSource='ds_name_rollup_for_pv', context={forceTimeChunkLock=true}}]
java.lang.RuntimeException: org.apache.druid.java.util.common.ISE: Failed to post task[AbstractTask{id='single_phase_sub_task_ds_name_rollup_for_pv_2020-05-29T07:53:14.337Z', groupId='index_parallel_ds_name_rollup_for_pv_2020-05-29T07:52:45.317Z', taskResource=TaskResource{availabilityGroup='single_phase_sub_task_ds_name_rollup_for_pv_2020-05-29T07:53:14.337Z', requiredCapacity=1}, dataSource='ds_name_rollup_for_pv', context={forceTimeChunkLock=true}}] with error[{"error":"Task[single_phase_sub_task_ds_name_rollup_for_pv_2020-05-29T07:53:14.337Z] already exists!"}].
	at org.apache.druid.client.indexing.HttpIndexingServiceClient.runTask(HttpIndexingServiceClient.java:134) ~[druid-server-0.16.1-incubating.jar:0.16.1-incubating]
	at org.apache.druid.indexing.common.task.batch.parallel.TaskMonitor.submitTask(TaskMonitor.java:268) ~[druid-indexing-service-0.16.1-incubating.jar:0.16.1-incubating]
	at org.apache.druid.indexing.common.task.batch.parallel.TaskMonitor.submit(TaskMonitor.java:225) ~[druid-indexing-service-0.16.1-incubating.jar:0.16.1-incubating]
	at org.apache.druid.indexing.common.task.batch.parallel.ParallelIndexPhaseRunner.submitNewTask(ParallelIndexPhaseRunner.java:230) ~[druid-indexing-service-0.16.1-incubating.jar:0.16.1-incubating]
	at org.apache.druid.indexing.common.task.batch.parallel.ParallelIndexPhaseRunner.run(ParallelIndexPhaseRunner.java:138) ~[druid-indexing-service-0.16.1-incubating.jar:0.16.1-incubating]
	at org.apache.druid.indexing.common.task.batch.parallel.ParallelIndexSupervisorTask.runNextPhase(ParallelIndexSupervisorTask.java:268) ~[druid-indexing-service-0.16.1-incubating.jar:0.16.1-incubating]
	at org.apache.druid.indexing.common.task.batch.parallel.ParallelIndexSupervisorTask.runSinglePhaseParallel(ParallelIndexSupervisorTask.java:448) ~[druid-indexing-service-0.16.1-incubating.jar:0.16.1-incubating]
	at org.apache.druid.indexing.common.task.batch.parallel.ParallelIndexSupervisorTask.runTask(ParallelIndexSupervisorTask.java:391) ~[druid-indexing-service-0.16.1-incubating.jar:0.16.1-incubating]
	at org.apache.druid.indexing.common.task.AbstractBatchIndexTask.run(AbstractBatchIndexTask.java:137) ~[druid-indexing-service-0.16.1-incubating.jar:0.16.1-incubating]
	at org.apache.druid.indexing.overlord.SingleTaskBackgroundRunner$SingleTaskBackgroundRunnerCallable.call(SingleTaskBackgroundRunner.java:419) [druid-indexing-service-0.16.1-incubating.jar:0.16.1-incubating]
	at org.apache.druid.indexing.overlord.SingleTaskBackgroundRunner$SingleTaskBackgroundRunnerCallable.call(SingleTaskBackgroundRunner.java:391) [druid-indexing-service-0.16.1-incubating.jar:0.16.1-incubating]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_232]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_232]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_232]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_232]
Caused by: org.apache.druid.java.util.common.ISE: Failed to post task[AbstractTask{id='single_phase_sub_task_ds_name_rollup_for_pv_2020-05-29T07:53:14.337Z', groupId='index_parallel_ds_name_rollup_for_pv_2020-05-29T07:52:45.317Z', taskResource=TaskResource{availabilityGroup='single_phase_sub_task_ds_name_rollup_for_pv_2020-05-29T07:53:14.337Z', requiredCapacity=1}, dataSource='ds_name_rollup_for_pv', context={forceTimeChunkLock=true}}] with error[{"error":"Task[single_phase_sub_task_ds_name_rollup_for_pv_2020-05-29T07:53:14.337Z] already exists!"}].
	at org.apache.druid.client.indexing.HttpIndexingServiceClient.runTask(HttpIndexingServiceClient.java:119) ~[druid-server-0.16.1-incubating.jar:0.16.1-incubating]
	... 14 more
2020-05-29T15:53:25,003 INFO [task-runner-0-priority-0] org.apache.druid.indexing.overlord.TaskRunnerUtils - Task [index_parallel_ds_name_rollup_for_pv_2020-05-29T07:52:45.317Z] status changed to [FAILED].
2020-05-29T15:53:25,010 INFO [task-runner-0-priority-0] org.apache.druid.indexing.worker.executor.ExecutorLifecycle - Task completed with status: {
  "id" : "index_parallel_ds_name_rollup_for_pv_2020-05-29T07:52:45.317Z",
  "status" : "FAILED",
  "duration" : 17567,
  "errorMsg" : "java.lang.RuntimeException: org.apache.druid.java.util.common.ISE: Failed to post task[AbstractTask{...",
  "location" : {
    "host" : null,
    "port" : -1,
    "tlsPort" : -1
  }
}
```

In my case, there is a daily scheduling script almost simultaneously submits 24 `index_parallel` tasks(one hour data per task), which aims to compact/create optimisation rollups for a specific datasource. And the `maxNumConcurrentSubTasks` property is set to 10 for every `index_parallel` task. Also found that each `index_parallel` task could submit 10 `single_phase_sub` tasks in about 1 second and currently we construct taskId for `single_phase_sub` task by appending a timestamp suffix. I think that explains the taskId conflict.

In local codebase, I enhance uniqueness of taskId of `single_phase_sub` task by appending a unique hash suffix(calculated from a string of timestamp+randomId) instead of just a timestamp suffix. And it works well.

In my opinion, the `index_parallel` type of task is the rising star among index type tasks and its parallel processing depends on `single_phase_sub` tasks, which means it may be very likely to happen that user simultaneously submit more than one `index_parallel` tasks for a specific datasource. So it's necessary to improve uniqueness of taskId of single_phase_sub` task. I will be glad to hear some feedback from you about this.

**Comment by jihoonson** [comment]
I think this was fixed in https://github.com/apache/druid/pull/8612. Could you please try out 0.17 or later versions?

**Comment by yuanlihan** [comment]
> I think this was fixed in #8612. Could you please try out 0.17 or later versions?

Thanks for pointing that out. It should have been fixed.

**Comment by jihoonson** [comment]
Thanks. Please feel free to reopen if it still happens.
